### ISSUE TITLE
Generic ClassVar Type failing

### ISSUE THREAD

**Issue by alanhdu** [python/mypy#5144]
With mypy `0.600` (and the latest master `eb1bb064d707ce05735ff6795df82126e75ea6ea`), running mypy on

```python
from typing import ClassVar, Type, Generic, TypeVar

_T = TypeVar('_T')

class Array(Generic[_T]):
    _length_: ClassVar[int]
    _type_: ClassVar[Type[_T]]

def generate() -> Type[Array[float]]:
    pass

generate()._type_(3)
```

fails with

```python
test.py:12: error: Too many arguments for "object"
```

instead of allowing this to construct a `float`.

The original use-case here is in the `ctypes` stubs, where mypy fails to derive the type of `(ctypes.c_float * 5)._type_` -- see:

https://github.com/python/typeshed/blob/9ec6d476c44729ea84e9cafde9c5a77610451b0c/stdlib/2and3/ctypes/__init__.pyi#L248-L250

**Comment by sutaner** [comment]
Maybe it has something to do with:

> Note that a ClassVar parameter cannot include any type variables, regardless of the level of nesting: ClassVar[T] and ClassVar[List[Set[T]]] are both invalid if T is a type variable.

See PEP 526 (https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations)

**Comment by ilevkivskyi** [comment]
One comment is now have good stubs and a plugin for `ctypes`, another comment is maybe this restriction in the PEP is to limiting, _accessing_ generic class variables when they are still generic should indeed be prohibited, but declaring it and accessing only on subclasses that has that variable bound should be OK (technically not 100% safe but but I think should be fine).

**Comment by ktbarrett** [comment]
I agree the PEP does seem unnecessarily limiting. If you want to use a type variable at runtime, the idiom might look like:

```python
class ListWithElementType(List[T], Generic[T]):
    element_type: ClassVar[Type[T]]
    # methods that use the element_type at runtime via type(self).element_type

class IntList(ListWithElementType[int]):
    element_type = int
```

`ListWithElementType` has an abstract attribute and trying to instantiate it should fail. Subclasses bind a value to the abstract attribute and wouldn't fail. An exception could be made specifically for ClassVar of TypeVar if it is abstract.

The following cases should fail. Maybe they were the cases they had in mind when writing that rule.

```python
class Bad(Generic[T]):
    thing: ClassVar[T] = some_func[T]()

Num = TypeVar("Num", bound=float)

class AlsoBad(Generic[Num]):
   thing: ClassVar[Num] = 0  # invalid until minimum bounds are supported (if ever)
```

**Comment by michaeloliverx** [comment]
> I agree the PEP does seem unnecessarily limiting. If you want to use a type variable at runtime, the idiom might look like:
> 
> ```python
> class ListWithElementType(List[T], Generic[T]):
>     element_type: ClassVar[Type[T]]
>     # methods that use the element_type at runtime via type(self).element_type
> 
> class IntList(ListWithElementType[int]):
>     element_type = int
> ```

This is exactly what I was looking to do, does the PEP need to be revised first for this to happen?

**Comment by embray** [comment]
Ditto--I'm not sure why this limitation was made in the first place.  Of course it could be violated by an assignment expression, but for an abstract base class it's not such an unusual thing to do.

**Comment by bedlamzd** [comment]
Here just to say that I've stumbled into this problem when defining abstract class, so I'd be happy if this restriction is lifted. Do you know if there is any issue/proposal in the python repo to revise related pep?

**Comment by Tinche** [comment]
I changed the attrs plugin recently to correctly generate the `__attrs_attrs__` property as a ClassVar without knowing Mypy doesn't support generic ClassVars. So now the natural protocol definition for an attrs class doesn't work, which means the signature of `fields()` (`(type[AttrsClass[T]] -> T`) cannot be correct and I'm a sad panda.

(The proposed protocol:
```python
_T = TypeVar("_T")

class AttrsClass(Protocol[_T]):
    __attrs_attrs__: ClassVar[_T]
```
)

**Comment by jchalupka-pan** [comment]
I believe that the reasoning given by @ktbarrett above makes sense. Anybody knows if there is any movement towards this? Any PEPs, discussions or anything?

**Comment by erictraut** [comment]
There has been some discussion in the [python/typing discussion forum](https://github.com/python/typing/discussions/1424), but there hasn't been any recent movement toward eliminating this limitation. If you're interested in championing such a change, I recommend participating in the python/typing forums.

**Comment by caniko** [comment]
> There has been some discussion in the https://github.com/python/typing/discussions/1424, but there hasn't been any recent movement toward eliminating this limitation.

@levkivskyi did [comment](https://github.com/python/mypy/issues/5144#issuecomment-439524567) on the matter earlier, and it does make sense. We need a PR to move forward.
